tribe-ext-pdf-tickets/index.php

```diff
@@ -0,0 +1,892 @@
+	 *                       setups).
+	 * @param $requested_url Our unmodified URL--without a trailing slash.
+	 *
+	 * @return bool|string
+	 */
+	public function make_non_trailing_slash_the_canonical( $redirect_url, $requested_url ) {
+		$pattern_wo_slash = sprintf( '/\/%s\/(%s)/', $this->get_download_base_slug(), $this->get_unique_id_regex() );
+
+		if ( preg_match( $pattern_wo_slash, $requested_url ) ) {
+			return false;
+		}
+
+		return $redirect_url;
+	}
+
+
+	/**
+	 * Flush WordPress rewrite rules when this extension is activated.
+	 *
+	 * @link https://codex.wordpress.org/Function_Reference/register_post_type#Flushing_Rewrite_on_Activation
+	 */
+	public function flush_rewrite_rules_on_activation() {
+		$this->add_pdf_file_rewrite_rules();
+
+		flush_rewrite_rules();
+	}
+
+	/**
+	 * Create PDF, save to server, and add to email queue.
+	 *
+	 * @param      $attendee_id ID of attendee ticket.
+	 * @param bool $email       Add PDF to email attachments array.
+	 *
+	 * @return bool
+	 */
+	public function do_upload_pdf( $attendee_id, $email = true ) {
+		$successful = false;
+
+		$ticket_provider_data = tribe_tickets_get_ticket_provider( $attendee_id );
+		$ticket_class         = $ticket_provider_data->className;
+		$ticket_instance      = $ticket_class::get_instance();
+
+		if ( empty( $ticket_instance ) ) {
+			return $successful;
+		}
+
+		// should only be one result
+		$event_ids = tribe_tickets_get_event_ids( $attendee_id );
+
+		if ( ! empty( $event_ids ) ) {
+			$event_id  = $event_ids[0];
+			$attendees = $ticket_instance->get_attendees_array( $event_id );
+		}
+
+		if (
+			empty( $attendees )
+			|| ! is_array( $attendees )
+		) {
+			return $successful;
+		}
+
+		$attendees_array = array();
+
+		foreach ( $attendees as $attendee ) {
+			if ( $attendee['attendee_id'] == $attendee_id ) {
+				$attendees_array[] = $attendee;
+			}
+		}
+
+		if ( empty( $attendees_array ) ) {
+			return $successful;
+		}
+
+		/**
+		 * Because $html is the full HTML DOM sent to the PDF generator, adding
+		 * anything to the beginning or the end would likely cause problems.
+		 *
+		 * If you want to alter what gets sent to the PDF generator, follow the
+		 * Themer's Guide for tickets/email.php or use that template file's
+		 * existing hooks.
+		 *
+		 * @link https://theeventscalendar.com/knowledgebase/themers-guide/#tickets
+		 */
+		$html = $ticket_instance->generate_tickets_email_content( $attendees_array );
+
+		if ( empty( $html ) ) {
+			return $successful;
+		}
+
+		$file_name = $this->get_pdf_path( $attendee_id );
+
+		if ( empty( $file_name ) ) {
+			return $successful;
+		}
+
+		if ( file_exists( $file_name ) ) {
+			$successful = true;
+		} else {
+			$successful = $this->output_pdf( $html, $file_name );
+		}
+
+		/**
+		 * Action hook after the PDF ticket gets created.
+		 *
+		 * Might be useful if you want the PDF file added to the Media Library
+		 * via wp_insert_attachment(), for example.
+		 *
+		 * @var $ticket_class
```

Comment:
Switch to `@param` when documenting hook args [(see here)](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters)